### PR TITLE
phase randomization score map correction

### PR DESCRIPTION
**Pull request by McHaillet**
This PR adds 'phase randomization', introduced by STOPGAP.

Explanation of what this is: The match_template command line got two additional arguments: one flag to activate this mode, another to set a seed for randomization. The seed is used to create a phase randomized version of the template, meaning the phases of the Fourier transform are randomly permuted (used often for FSC calculations). This second template is like a random noise object and is matched simultaneously with the original template. The score map of the noisy version is subtracted from the real one.

With the particle diameter updated and ctf input reworked, this is the last thing I really want to add in functionality-wise.

Once this one is merged, I would like make a PR to throw the code through ruff and maybe set up a pre-commit thingy.

### CODE REVIEW

File: src/pytom_tm/entry_points.py [hunk 1]
```diff
@@ -756,6 +756,26 @@ def match_template(argv=None):
...
+        required=False,
+        help="Run template matching simultaneously with a phase randomized version of "
+             "the template, and subtract this 'noise' map from the final score map. "
+             "For this method please see STOPGAP as a reference: "
+             "https://doi.org/10.1107/S205979832400295X ."
+    )
+    additional_group.add_argument(
+        "--rng-seed",
+        action=LargerThanZero,
+        default=321,
```
Comment (sroet) [inline src/pytom_tm/entry_points.py:774]: I wonder if it is best to have no default for this option (generate a random seed every time this is ran)

File: src/pytom_tm/matching.py [hunk 2]
```diff
@@ -101,7 +116,8 @@ def __init__(
         - pyTME: https://github.com/KosinskiLab/pyTME
 
         The precalculation of conjugated FTs of the tomo was (AFAIK) introduced
-        by STOPGAP!
+        by STOPGAP! Also, they introduced simultaneous matching with a phase randomized
+        version of the template.
```
Comment (sroet) [inline src/pytom_tm/matching.py:120]: maybe also add doi link here?
```suggestion
        version of the template. https://doi.org/10.1107/S205979832400295X
```

File: src/pytom_tm/matching.py [hunk 3]
```diff
@@ -262,6 +297,51 @@ def run(self) -> tuple[npt.NDArray[float], npt.NDArray[float], dict]:
...
+                # Paste in center
+                self.plan.template_padded[pad_index] = self.plan.template
+
+                # Fast local correlation function between volume and template, norm is the standard deviation at each
+                # point in the volume in the masked area
+                self.plan.ccc_map = (
+                        irfftn(self.plan.volume_rft_conj * rfftn(self.plan.template_padded),
+                               s=self.plan.template_padded.shape)
+                        / self.plan.std_volume
+                )
```
Comment (sroet) [inline src/pytom_tm/matching.py:330]: this part is very similar to 257-285 and should in my opinion be split out into a separate function

File: src/pytom_tm/template.py [hunk 4]
```diff
@@ -103,3 +108,43 @@ def generate_template_from_map(
...
+    relevant_freqs = grid <= 1  # permute only up to Nyquist
+    noise = np.zeros_like(phase)
+    rng = np.random.default_rng(seed)
+    noise[relevant_freqs] = rng.permutation(phase[relevant_freqs])
+
+    # construct the new template
+    noise = np.reshape(noise, amplitude.shape)
+    result = np.fft.irfftn(
+        amplitude * np.exp(1j * noise), s=template.shape
+    )
```
Comment (sroet) [inline src/pytom_tm/template.py:149]: I think this should be using the imported scipy function to make sure we don't run into issues with #171 
```suggestion
    result = irfftn(
        amplitude * np.exp(1j * noise), s=template.shape
    )
```

**Review by sroet — CHANGES_REQUESTED** [review]
general comment:
- ~min coverage should be increased to 86~ already done in #184 

Other than that, one default question, one proposed refactor, one doc suggestion, one code consistency thing

> Once this one is merged, I would like make a PR to throw the code through ruff and maybe set up a pre-commit thingy.

Sounds good, I can also look into that if you don't have time for that (isn't necessary 'publishable' :wink:  ), just tag me into an issue

File: src/pytom_tm/entry_points.py (hunk 1 above)
Comment (McHaillet) [reply src/pytom_tm/entry_points.py:774]: probably, can add that

File: src/pytom_tm/matching.py (hunk 3 above)
Comment (McHaillet) [reply src/pytom_tm/matching.py]: Did it!

**Comment by McHaillet** [comment]
> Sounds good, I can also look into that if you don't have time for that (isn't necessary 'publishable' 😉 ), just tag me into an issue

Also fine, will respond in the PR I made for it!

**Review by sroet — APPROVED** [review]
LGTM, feel free to merge

**Pull request merged**